clang/test/CodeGenSYCL/intel-fpga-ivdep-global.cpp

```diff
@@ -0,0 +1,124 @@
+    a[i] = 0;
+    // CHECK:  %{{[0-9a-z]+}} = getelementptr inbounds [10 x i32], [10 x i32]* %[[ARRAY_B]], i64 0, i64 %{{[0-9a-z]+}}, !llvm.index.group ![[IDX_GROUP_B_SAFELEN:[0-9]+]]
+    b[i] = 0;
+    // CHECK: br label %for.cond, !llvm.loop ![[MD_LOOP_SAFELEN:[0-9]+]]
+  }
+}
+
+// Global ivdep, albeit conflicting safelens - annotate all GEPs
+//
+// CHECK: define spir_func void @_Z{{[0-9]+}}ivdep_conflicting_safelenv()
+void ivdep_conflicting_safelen() {
+  // CHECK: %[[ARRAY_A:[0-9a-z]+]] = alloca [10 x i32]
+  int a[10];
+  // CHECK: %[[ARRAY_B:[0-9a-z]+]] = alloca [10 x i32]
+  int b[10];
+  [[intelfpga::ivdep(5)]]
+  [[intelfpga::ivdep(4)]]
+  for (int i = 0; i != 10; ++i) {
+    // CHECK:  %{{[0-9a-z]+}} = getelementptr inbounds [10 x i32], [10 x i32]* %[[ARRAY_A]], i64 0, i64 %{{[0-9a-z]+}}, !llvm.index.group ![[IDX_GROUP_A_CONFL_SAFELEN:[0-9]+]]
+    a[i] = 0;
+    // CHECK:  %{{[0-9a-z]+}} = getelementptr inbounds [10 x i32], [10 x i32]* %[[ARRAY_B]], i64 0, i64 %{{[0-9a-z]+}}, !llvm.index.group ![[IDX_GROUP_B_CONFL_SAFELEN:[0-9]+]]
+    b[i] = 0;
+    // CHECK: br label %for.cond, !llvm.loop ![[MD_LOOP_CONFL_SAFELEN:[0-9]+]]
+  }
+}
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+  kernel_single_task<class kernel_function>([]() {
+    ivdep_no_param();
+    ivdep_no_param_multiple_geps();
+    ivdep_safelen();
+    ivdep_conflicting_safelen();
+  });
+  return 0;
+}
+
+/// Global ivdep w/o safelen specified
+/// All arrays have the same INF safelen - put access groups into the same parallel_access_indices metadata
+//
+// CHECK-DAG: ![[IDX_GROUP_A_NO_PARAM]] = distinct !{}
+// CHECK-DAG: ![[IDX_GROUP_B_NO_PARAM]] = distinct !{}
+// CHECK-DAG: ![[MD_LOOP_NO_PARAM]] = distinct !{![[MD_LOOP_NO_PARAM]], ![[IVDEP_NO_PARAM:[0-9]+]]}
+// CHECK-DAG: ![[IVDEP_NO_PARAM]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_A_NO_PARAM]], ![[IDX_GROUP_B_NO_PARAM]]}
+//
+// CHECK-DAG: ![[IDX_GROUP_A_MUL_GEPS]] = distinct !{}
+// CHECK-DAG: ![[IDX_GROUP_B_MUL_GEPS]] = distinct !{}
+// CHECK-DAG: ![[MD_LOOP_MUL_GEPS]] = distinct !{![[MD_LOOP_MUL_GEPS]], ![[IVDEP_MUL_GEPS:[0-9]+]]}
+// CHECK-DAG: ![[IVDEP_MUL_GEPS]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_A_MUL_GEPS]], ![[IDX_GROUP_B_MUL_GEPS]]}
+
+/// Global ivdep w/ safelen specified
+/// All arrays share the same safelen - put index groups into the same parallel_access_indices MD node
+//
+// CHECK-DAG: ![[IDX_GROUP_A_SAFELEN]] = distinct !{}
+// CHECK-DAG: ![[IDX_GROUP_B_SAFELEN]] = distinct !{}
+// CHECK-DAG: ![[MD_LOOP_SAFELEN]] = distinct !{![[MD_LOOP_SAFELEN]], ![[IVDEP_SAFELEN:[0-9]+]]}
+// CHECK-DAG: ![[IVDEP_SAFELEN]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_A_SAFELEN]], ![[IDX_GROUP_B_SAFELEN]], i32 5}
```

Comment:
That's looking wrong. If we have to chose between safelen = 4 and safelen = 5, I would prefer to go the safest path and pick the lowest of them.

Reply:
That is not how this was specified to me.  @AGindinson  can you please comment?

Reply:
Sorry, I missed the test case. The comment should be applied to the check on string below. But it remains the same.

Reply:
@MrSidims, the greatest safelen is meant to be picked (just like INF safelen overrides a specified finite safelen). In terms of this being unsafe, IVDep is merely a hint to the backend - it does request ignoring vector dependencies on a specified safelen, but rather **to the extent possible**.

Reply:
That remains unclear for me. I see your point, but I also can imagine a following use case:
```
int a[10];
int b[10];
int c[10];
[[intelfpga::ivdep]] // the loop caries no dependencies in general
[[intelfpga::ivdep(c, 1)]] // the loop caries dependencies for an array c
for (;;) {
  a[i] = ....; // do some optimizations
  b[i] = ....; // do some optimizations
  c[i] = ....; // be careful there
}
```

Does it make sense?

Reply:
In that case, you'd be better off applying `intelfpga::ivdep(a)` and `intelfpga::ivdep(b)` w/o mentioning `c`, as ivdep(1) is essentially meaningless.
Moreover, leaving just the global ivdep wouldn't have any severe consequences - if the optimization cannot be performed for `c` due to the cross-iteration dependencies, it simply won't happen. Again, it all comes down to ivdep being a mere hint that allows the user to fine-tune their code, but does not force any unwanted behavior.